docs/architecture/adr-036-blockchain-reactor-refactor.md

```diff
@@ -0,0 +1,406 @@
+current design and the implementation, and the problem with writing tests.
+More precisely, the current architecture consists of dozens of routines and it is tightly depending on the `Switch`, making writing unit tests almost impossible. Current tests require setting up complex dependency graphs and dealing with concurrency. Note that having dozens of routines is in this case overkill as most of the time routines sits idle waiting for something to happen (message to arrive or timeout to expire). Due to dependency on the `Switch`, 
+testing relatively complex network scenarios and failures (for example adding and removing peers) is very complex tasks and frequently lead to complex tests with not deterministic behavior ([#3400]).   
+
+This resulted in several issues (some are closed and some are still open): 
+[#3400], [#2897], [#2896], [#2699], [#2888], [#2457], [#2622], [#2026].  
+
+## Decision
+
+To remedy these issues we plan a major refactor of the blockchain reactor. The proposed architecture is largely inspired by ADR-30 and is presented 
+on the following diagram:
+![Blockchain Reactor Refactor Diagram](img/bc-reactor-refactor.png)
+
+We suggest a concurrency architecture where the core algorithm (we call it `Controller`) is extracted into a finite state machine. 
+The active routine of the reactor is called `Executor` and is responsible for receiving and sending messages from/to peers and triggering timeouts.
+What messages should be sent and timeouts triggered is defined mostly by the `Controller`. The exception is peer heartbeat mechanism which is `Executor`
+responsibility. The heartbeat mechanism consists of periodically sending `StatusRequest` messages to peers and signalling to `Executor` in case a peer does not respond 
+timely with the corresponding `StatusReport` message that a peer is not responsive so it should be removed from the peer list.    
+This simpler architecture is easier to understand and makes writing of unit tests simpler tasks as most of the critical
+logic is part of the `Controller` function. 
+
+### Implementation changes
+
+We assume the following system model for "fast sync" protocol: 
+
+* a node is connected to a random subset of all nodes that represents its peer set. Some nodes are correct and some might be faulty. We don't make assumptions about ratio of  
+  faulty nodes, i.e., it is possible that all nodes in some peer set are faulty. 
+* we assume that communication between correct nodes is synchronous, i.e., if a correct node `p` sends a message `m` to a correct node `q` at time `t`, then `q` will receive
+  message the latest at time `t+Delta` where `Delta` is a system parameter that is known. `Delta` is normally order of magnitude higher than the real communication delay 
+  (maximum) between correct nodes. Therefore if a correct node `p` sends a request message to a correct node `p` at time `t` and there is no the corresponding reply at time
```

Comment:
`Therefore if a correct node p sends a request message to a correct node p` to `Therefore if a correct node p sends a request message to a correct node q`

Reply:
```suggestion
  (maximum) between correct nodes. Therefore if a correct node `p` sends a request message to a correct node `q` at time `t` and there is no the corresponding reply at time
```